lightning/src/ln/channelmanager.rs

```diff
@@ -3305,6 +3327,13 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					}
 					if payment.get().remaining_parts() == 0 {
 						all_paths_failed = true;
+						if payment.get().retries_exceeded() {
+							full_failure_ev = Some(events::Event::PaymentFailed {
+								payment_id,
+								payment_hash: payment.get().payment_hash().expect("RetriesExceeded PendintOutboundPayments always have a payment hash set"),
+							});
```

Comment:
I think you can avoid creating this here by conditioning on `all_paths_failed` when needing to add the event later.

Reply:
We also have to test that `retries_exceeded` is set, so I don't think we can after here where we drop the payment object itself.

Reply:
Ah, I was thinking you could still create the event without the entry since the payment hash is available below. But I guess you would need another variable regardless, so probably not worth it.